### ISSUE TITLE
unit testing with axios

### ISSUE THREAD

**Issue by gaderrick** [susanBuck/e28-fall19#158]
For P4 (a recipe project), I'm trying to write a unit for the component that loads a single recipe onto the page from an axios api call. In my .vue page the following is defined:
```javascript
export default {
  name: 'RecipePage',
  props: ['id'],
  ...
}
```
The component is expecting a single data element (id) to be passed in which I set up in the file RecipePage.spec.js as follows:
```javascript
describe('RecipePage.vue', () => {
    it('shows recipe contents', () => {
        let id = 1
        const wrapper = shallowMount(RecipePage, {
            propsData: { id }
        })
        expect(wrapper.text()).to.include("Cream Ale");
    })
})
```
When I run the test, I get the following error:
```
 MOCHA  Testing...

  RecipePage.vue
    1) shows recipe contents

  0 passing (102ms)
  1 failing

  1) RecipePage.vue
       shows recipe contents:
     AssertionError: expected '' to include 'Cream Ale'
      at Context.it (dist\js\webpack:\tests\unit\RecipePage.spec.js:11:1)
```
My assumption (based on the AssertionError) is that it has something to do with the axios call not making the api call and returning/processing it in the HTML, but I'm not sure.

Any ideas/suggestions would be helpful.

**Comment by susanBuck** [comment]
Hi @gaderrick -

Unit tests are designed to test individual units, in isolation.

Given this, we don't really want to test the Axios functionality within the component.

But how can we test the component if it depends on an Axios call?

The answer is to mock the Axios call. In other words, fake the call, returning data we know is correct. That way we can assume that, if given the correct data, the component is functioning as expected.

And then our E2E tests will fill the gaps and confirm that our component *and* its nested Axios call are successfully working together as they should. This is where the E2E tests excel - because they test all the units together, end to end.

__As for how to mock Axios calls in your unit tests...__

One solution is to use this library: [Moxios](https://github.com/axios/moxios).

Because I haven't provided an example of this, I'm not expecting this in Project 4, which is part of the reason I'm only requiring you unit test at least 1 component (with the assumption you'll have at least 1 component that does not require axios :)

__In short - the easiest solution to meet Project 4 requirements is to unit test another component, and leave E2E tests for your Axios-dependent components.__

All that being said, beyond project requirements, I do want to provide an example of using Moxios for those that are curious. I've got a packed schedule this week, but I'll make a post next week with an example. 

Alternatively, if someone wants to take a crack at integrating Moxios into a test in ZipFoods and they share their solution before I get to it next week - go for it! : )

**Comment by susanBuck** [comment]
Related/for reference: [localStorage is not defined error in unit test #157](#157)

**Comment by susanBuck** [comment]
Also related: https://vue-test-utils.vuejs.org/guides/testing-async-components.html#asynchronous-behavior-outside-of-vue

**Comment by gaderrick** [comment]
@susanBuck thanks for the Moxios pointer, I'll look at it out of curiousity if nothing else. I ended up modifying your example code to redo my app a little bit to use computed properties. Then I just overroad the computed property with the data that would have come back from the api. Seems to be working.

**Comment by susanBuck** [comment]
Good call on harnessing a computed property. It slipped my mind that I had showed that technique for bypassing the Vuex store in the unit tests, so I'm glad you brought it up again. : )

```js
it('shows all the products', () => {

        const wrapper = mount(ProductsPage, {
            // We can pass mount a `computed` option and within it we can overwrite
            // any of our componenent's computed properties.
            // This allows us to bypass the Vuex store for testing purposes
            // For alternative ways of dealing with Vuex when testing components, see the following:
            // https://vue-test-utils.vuejs.org/guides/#testing-vuex-in-components
            // https://lmiller1990.github.io/vue-testing-handbook/testing-vuex.html
            computed: {
                products: function () {
                    return [product]
                }
            },
            stubs: {
                RouterLink: RouterLinkStub
            }
        })

        expect(wrapper.text()).to.include('Products');
        expect(wrapper.text()).to.include('Driscoll’s Strawberries');
    })
```
